### ISSUE TITLE
[Data] iter_batches(local_shuffle_buffer_size=...) yields short batches mid-stream, breaking DDP step parity

### ISSUE THREAD

**Issue by p-crsx** [ray-project/ray#66085]
### What happened + What you expected to happen

When local_shuffle_buffer_size is set, iter_batches / iter_torch_batches can emit batches smaller than batch_size in the middle of the stream, while many rows are still queued.

Both docstrings promise this only happens at the end:

> batch_size: … The final batch may include fewer than batch_size rows if drop_last is False.

> local_shuffle_buffer_size: … the minimum number of rows that must be in the local in-memory shuffle buffer in order to yield a batch. When there are no more rows to add to the buffer, the remaining rows in the buffer will be drained.

ray.train.DataConfig shards with streaming_split(world_size, equal=True), which guarantees every worker the same number of rows. It does not guarantee the same number of blocks — the splitter dispatches whole blocks greedily to whichever output is behind, and only equalizes rows at the end. Combined with the above, two workers holding identical row counts produce different batch counts, so they run a different number of training steps.

### Versions / Dependencies

ray       2.56.0
python    3.11.7
pyarrow   23.0.1
pandas    2.3.3
numpy     1.26.4
platform  Linux 6.18 x86_64, glibc 2.28

### Reproduction script

"""Ray Data: `iter_batches(local_shuffle_buffer_size=...)` yields short batches mid-stream.

Part 1 shows the contract violation on a single dataset (40 rows, no split).
Part 2 shows the consequence for distributed training: `streaming_split(equal=True)`
gives shards equal rows but unequal *batch* counts, which deadlocks DDP.
"""

```
import numpy as np
import pandas as pd
import ray

BATCH_SIZE = 8
SHUFFLE_BUFFER = 4


def dataset(block_sizes):
    """One block per entry, so block boundaries are exactly `block_sizes`."""
    offset, frames = 0, []
    for n in block_sizes:
        frames.append(pd.DataFrame({"value": np.arange(offset, offset + n)}))
        offset += n
    return ray.data.from_pandas(frames)


def part1():
    print("=" * 70)
    print("Part 1: short batch mid-stream on a single dataset (40 rows, 2 blocks)")
    print("=" * 70)
    ds = dataset([30, 10])
    for buffer_size in (None, SHUFFLE_BUFFER):
        sizes = [
            len(b["value"])
            for b in ds.iter_batches(
                batch_size=BATCH_SIZE, local_shuffle_buffer_size=buffer_size
            )
        ]
        note = ""
        if any(s != BATCH_SIZE for s in sizes[:-1]):
            note = "   <-- short batch BEFORE the last one"
        print(f"  local_shuffle_buffer_size={str(buffer_size):4}  "
              f"{len(sizes)} batches  sizes={sizes}{note}")
    print("\n  Expected: [8, 8, 8, 8, 8] in both cases. Only the *final* batch is documented\n"
          "  as possibly short, and 40 rows divide evenly into 5 batches of 8 anyway.\n")


@ray.remote
def drain(shard, buffer_size):
    return [
        len(b["value"])
        for b in shard.iter_batches(
            batch_size=BATCH_SIZE, local_shuffle_buffer_size=buffer_size
        )
    ]


def part2():
    print("=" * 70)
    print("Part 2: streaming_split(equal=True) -> equal rows, unequal batch counts")
    print("=" * 70)
    ds = dataset([30, 10, 10, 30])  # 80 rows, 4 blocks
    for buffer_size in (None, SHUFFLE_BUFFER):
        out = ray.get([drain.remote(s, buffer_size)
                       for s in ds.streaming_split(2, equal=True)])
        rows = [sum(s) for s in out]
        counts = [len(s) for s in out]
        print(f"  local_shuffle_buffer_size={str(buffer_size):4}")
        for i, sizes in enumerate(out):
            print(f"      shard {i}: {len(sizes)} batches  sizes={sizes}  rows={sum(sizes)}")
        print(f"      rows per shard    {rows}  (equal, as documented)")
        print(f"      batches per shard {counts}"
              f"{'  <-- DIFFER' if len(set(counts)) != 1 else ''}")
    print("\n  Under DDP the shard with fewer batches leaves the training loop while the\n"
          "  other is still in a backward pass, and the gradient all-reduce never completes.\n")


if __name__ == "__main__":
    ray.init(log_to_driver=False)
    part1()
    part2()

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.
